File: results/dnd-rest-benchmark/lifecycle-runs/20260730T072747Z_pi_kimi-k2p7-code_go-stdlib/dice.go
package main

import (
	"encoding/json"
	"net/http"
	"regexp"
	"strconv"
)

// diceRe matches dice expressions such as "2d6", "1d20+3", and "3d8-1".
// The capturing groups are: count, sides, optional sign, and optional modifier.
var diceRe = regexp.MustCompile(`^(\d+)d(\d+)(?:([+-])(\d+))?$`)

type diceStatsRequest struct {
	Expression string `json:"expression"`
}

type diceStatsResponse struct {
	DiceCount int     `json:"dice_count"`
	Sides     int     `json:"sides"`
	Modifier  int     `json:"modifier"`
	Min       int     `json:"min"`
	Max       int     `json:"max"`
	Average   float64 `json:"average"`
}

// diceStatsHandler parses a dice expression and returns deterministic statistics.
// It does not roll dice; it computes the range and average for the expression.
func diceStatsHandler(w http.ResponseWriter, r *http.Request) {
	var req diceStatsRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	matches := diceRe.FindStringSubmatch(req.Expression)
	if matches == nil {
		writeError(w, http.StatusBadRequest, "invalid expression")
		return
	}

	count, _ := strconv.Atoi(matches[1])
	sides, _ := strconv.Atoi(matches[2])
	modifier := 0
	if matches[3] != "" && matches[4] != "" {
		m, _ := strconv.Atoi(matches[4])
		if matches[3] == "-" {
			m = -m
		}
		modifier = m
	}

	if count <= 0 || sides <= 0 {
		writeError(w, http.StatusBadRequest, "invalid expression")
		return
	}

	min := count + modifier
	max := count*sides + modifier
	average := float64(count*(1+sides)+2*modifier) / 2.0

	writeJSON(w, http.StatusOK, diceStatsResponse{
		DiceCount: count,
		Sides:     sides,
		Modifier:  modifier,
		Min:       min,
		Max:       max,
		Average:   average,
	})
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T081050Z_pi_kimi-k2p7-code_php-symfony/src/Domain/Dice.php
<?php

declare(strict_types=1);

namespace App\Domain;

use App\Util\Numbers;

/**
 * Parse and analyse simple D&D dice expressions such as "2d6+3".
 */
final class Dice
{
    /**
     * Parse a dice expression of the form XdY[+|-Z].
     *
     * Returns null when the expression does not match the supported grammar.
     */
    public static function parse(string $expression): ?array
    {
        if (!preg_match('/^(?P<count>0*[1-9]\d*)d(?P<sides>0*[1-9]\d*)(?P<modifier>[+-]\d+)?$/', trim($expression), $matches)) {
            return null;
        }

        return [
            'count' => (int) $matches['count'],
            'sides' => (int) $matches['sides'],
            'modifier' => isset($matches['modifier']) ? (int) $matches['modifier'] : 0,
        ];
    }

    /**
     * Compute deterministic statistics for a parsed dice expression.
     */
    public static function stats(int $count, int $sides, int $modifier): array
    {
        return [
            'dice_count' => $count,
            'sides' => $sides,
            'modifier' => $modifier,
            'min' => $count + $modifier,
            'max' => $count * $sides + $modifier,
            'average' => Numbers::toIntegerWhenWhole($count * ($sides + 1) / 2 + $modifier),
        ];
    }
}
